neps/nep-0483.md

```diff
@@ -0,0 +1,83 @@
+---
+NEP: 0461
+Title: Prefix Tag for Signed Messages
+Author: Evgeny Kapun <eatmore@near.org>, Gallardo Guillermo <guillermo@near.org>, Firat Sertgoz <firat@near.org>
+DiscussionsTo: https://github.com/nearprotocol/neps/pull/461
+Status: Review
+Type: Protocol
+Category: Runtime
+Version: 1.0.0
+Created: 14-Feb-2023
+Updated: 14-Feb-2023
+---
+
+## Summary
+
+A proposal to include a standardized prefix in all standardized signable messages, to distinguish between (1) actionable messages (i.e. those that are validated on chain), (2) non-actionable messages (i.e. those that are used off-chain), and (3) [classic transactions](https://nomicon.io/RuntimeSpec/Transactions).
+
+Particularly, we propose to prepend an explicit prefix to the structure being signed using the following rules:
+
+| Prefix | Message |
+| - | - |
+| $< 2^{30}$ | [Classic Transactions](https://nomicon.io/RuntimeSpec/Transactions) |  
+| $2^{30}$ + NEP-number  | On-chain Actionable Message, e.g. [`SignedDelegateAction`](https://github.com/near/NEPs/blob/master/neps/nep-0366.md) |
+| $2^{31}$ + NEP-number  | Off-chain Message, e.g. [`signMessage`](https://github.com/near/NEPs/pull/413) |
+
+## Motivation
+
+As the ecosystem grows, more standards will appear, enabling users to sign messages that are not transactions (e.g. [NEP413](https://github.com/near/NEPs/pull/413/), and [NEP366](https://github.com/near/NEPs/blob/master/neps/nep-0366.md)). If such messages collide with transactions, a malicious app could ask users to sign them, only to later relay them as valid transactions to the network.
+
+## Rationale and alternatives
+
+In NEAR, users sign transactions that are later relayed to the network. At the moment of being relayed, the transactions are simple streams of bytes that a node then parses into a [valid transaction](https://nomicon.io/RuntimeSpec/Transactions).
+
+As the protocol expands, more messages will be added for users to sign. We want users to be able to use the protocol safely, making sure that all those signed messages cannot be interpreted as a valid transaction.
+
+In practice, what users sign is the `sha256` hash of a [borsh-serialized](https://borsh.io) structure. This is:
+
+```ts
+sha256.hash(Borsh.serialize(MessageStructure))
+```
+
+Including a large `u32` prefix allows to easily distinguish transactions from messages, ensuring that they never overlap.
+
+```ts
+sha256.hash(Borsh.serialize<u32>(large-number) + Borsh.serialize(MessageStructure))
```

Comment:
`original comment by` @mfornet

Before reading the suggestion, I thought about the user signing a message of this type:

```suggestion
Borsh.serialize<u32>(large-number) +  sha256.hash(Borsh.serialize(MessageStructure))
```
It is 4 bytes longer, but the advantage is that the wallet (signing device) can tell apart the target (i.e transaction/on-chain/off-chain) by just looking at this payload, rather than requiring the whole message. It is probably not safe anyway to sign a payload without seeing the message, but it might make sense for low-end devices (like some hardware wallets) where having the whole message is expensive or not possible.

Reply:
@mfornet wouldn't this have the problem that `sha256(transaction)` could collide with `borsh(prefix) + sha256(MessageStruture)`? If all transactions are of the form `sha256(prefix + transaction)` then by definition there cannot be a collision.